### PR TITLE
BI-8162 Define ordering for comparators download links

### PR DESCRIPTION
**Pull request by kinpang-CH**
- Add a new header to all triggers, orderNumber which represents its ordering number.
- Add new field to PublisherResourceRequest representing the orderNumber field.
- Add new class EmailLinksSorter to order a list of PublisherResourceRequest.
- Change SendEmailRoute to call above bean for sorting.

[BI-8162 (JIRA Ticket)](https://companieshouse.atlassian.net/browse/BI-8162)

### CODE REVIEW

**Review by phudson-ch — CHANGES_REQUESTED** [review]
1. Setting link presentation order at the level of trigger route definition results in dependency between trigger routes and has resulted in extra method parameters in places. e.g. add or remove trigger route or a change of order may result in changes to one or more routes. Consider that the ordering is a presentation issue that needs to be dealt with after the list of email links has been created.  As such the bean doing the ordering needs to know which email and what the expected link ordering should be. i.e., that component could be configured from properties e.g.,

comparison-email.<email id 1>.name=Company Profile
comparison_email.<email id 1>.<link id 1>.rank=1
comparison_email.<email id 1>.<link id 2>.rank=3
comparison_email.<email id 1>.<link id 3>.rank=2
...

Consider providing email identity to EmailAggregationStrategy->ResourceLinksWrapper and providing link identity to ResourceLink; that would enable an external email / link ordering scheme to be applied.

2. EmailLinksSorter contains conditional logic and email specific code: consider that 1 above could mitigate that by applying an external ordering scheme; or alternatively ResourceLinksWrapper could be preconfigured with an ordering scheme that would eliminate the EmailLinksSorter bean from the SendEmailRoute.

**Comment by kinpang-CH** [comment]
> 1. Setting link presentation order at the level of trigger route definition results in dependency between trigger routes and has resulted in extra method parameters in places. e.g. add or remove trigger route or a change of order may result in changes to one or more routes. Consider that the ordering is a presentation issue that needs to be dealt with after the list of email links has been created.  As such the bean doing the ordering needs to know which email and what the expected link ordering should be. i.e., that component could be configured from properties e.g.,
> 
> comparison-email.<email id 1>.name=Company Profile
> comparison_email.<email id 1>.<link id 1>.rank=1
> comparison_email.<email id 1>.<link id 2>.rank=3
> comparison_email.<email id 1>.<link id 3>.rank=2
> ...
> 
> Consider providing email identity to EmailAggregationStrategy->ResourceLinksWrapper and providing link identity to ResourceLink; that would enable an external email / link ordering scheme to be applied.
> 
> 1. EmailLinksSorter contains conditional logic and email specific code: consider that 1 above could mitigate that by applying an external ordering scheme; or alternatively ResourceLinksWrapper could be preconfigured with an ordering scheme that would eliminate the EmailLinksSorter bean from the SendEmailRoute.

- Tackled as part of our refactoring on the solution, seen in latest commits.

**Comment by kinpang-CH** [comment]
> 1. Setting link presentation order at the level of trigger route definition results in dependency between trigger routes and has resulted in extra method parameters in places. e.g. add or remove trigger route or a change of order may result in changes to one or more routes. Consider that the ordering is a presentation issue that needs to be dealt with after the list of email links has been created.  As such the bean doing the ordering needs to know which email and what the expected link ordering should be. i.e., that component could be configured from properties e.g.,
> 
> comparison-email.<email id 1>.name=Company Profile
> comparison_email.<email id 1>.<link id 1>.rank=1
> comparison_email.<email id 1>.<link id 2>.rank=3
> comparison_email.<email id 1>.<link id 3>.rank=2
> ...
> 
> Consider providing email identity to EmailAggregationStrategy->ResourceLinksWrapper and providing link identity to ResourceLink; that would enable an external email / link ordering scheme to be applied.
> 
> 1. EmailLinksSorter contains conditional logic and email specific code: consider that 1 above could mitigate that by applying an external ordering scheme; or alternatively ResourceLinksWrapper could be preconfigured with an ordering scheme that would eliminate the EmailLinksSorter bean from the SendEmailRoute.

- Tackled as part of our refactoring on the solution, seen in latest commits.

File: src/main/java/uk/gov/companieshouse/reconciliation/config/AggregationHandler.java [hunk 1]
```diff
@@ -1,12 +1,16 @@
 package uk.gov.companieshouse.reconciliation.config;
 
+import org.springframework.validation.annotation.Validated;
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/reconciliation/config/AggregationHandler.java:3]: Please remove this unused import.

File: src/main/java/uk/gov/companieshouse/reconciliation/function/email/PublisherResourceRequest.java [hunk 2]
```diff
@@ -55,16 +57,22 @@ public String getDescription() {
...
+        return linkId;
+    }
+
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         PublisherResourceRequest that = (PublisherResourceRequest) o;
-        return expirationTimeInMillis == that.expirationTimeInMillis && Objects.equals(objectKey, that.objectKey) && Objects.equals(uploaderEndpoint, that.uploaderEndpoint) && Objects.equals(presignerEndpoint, that.presignerEndpoint) && Objects.equals(resourceDescription, that.resourceDescription) && Objects.equals(comparisonGroup, that.comparisonGroup);
+        return expirationTimeInMillis == that.expirationTimeInMillis && Objects.equals(objectKey, that.objectKey) && Objects.equals(uploaderEndpoint, that.uploaderEndpoint) && Objects.equals(presignerEndpoint, that.presignerEndpoint) && Objects.equals(resourceDescription, that.resourceDescription) && Arrays.equals(results, that.results) && Objects.equals(comparisonGroup, that.comparisonGroup) && Objects.equals(linkId, that.linkId);
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/reconciliation/function/email/PublisherResourceRequest.java:69]: Please remove results from the equals expression; this violates the design contract of equals/hashCode and will result in a performance hit at runtime

File: src/main/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategy.java [hunk 3]
```diff
@@ -28,30 +41,53 @@
...
     @Override
     public Exchange aggregate(Exchange oldExchange, Exchange newExchange) {
-        //on first invocation, oldExchange will be null as this is the first message that we are aggregating
-        Exchange exchange = Optional.ofNullable(oldExchange).orElse(newExchange);
-        ResourceLinksWrapper downloadLinks = createOrGetResourceLinks(exchange);
         Optional<String> linkReference = header(newExchange, LINK_REFERENCE_HEADER);
-        if (linkReference.isPresent()) {
-            downloadLinks.addDownloadLink(linkReference.get(), header(newExchange, LINK_DESCRIPTION_HEADER).orElse(null));
-        } else {
+        if (! linkReference.isPresent()) {
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategy.java:52]: Conflicts with error handling work

File: src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLink.java [hunk 4]
```diff
@@ -1,18 +1,29 @@
 package uk.gov.companieshouse.reconciliation.model;
 
+import java.util.Objects;
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLink.java:3]: Please remove this unused import.

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/reconciliation/config/AggregationHandler.java (hunk 1 above)
Comment (phudson-ch) [reply src/main/java/uk/gov/companieshouse/reconciliation/config/AggregationHandler.java]: Done

File: src/main/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategy.java (hunk 3 above)
Comment (phudson-ch) [reply src/main/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategy.java:59]: Resolved

File: src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLink.java (hunk 4 above)
Comment (phudson-ch) [reply src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLink.java]: Done

File: src/main/java/uk/gov/companieshouse/reconciliation/function/email/PublisherResourceRequest.java (hunk 2 above)
Comment (phudson-ch) [reply src/main/java/uk/gov/companieshouse/reconciliation/function/email/PublisherResourceRequest.java]: Results removed from equals expression.

**Review by phudson-ch — APPROVED** [review]

**Review by cgregoryCH — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.companieshouse.reconciliation.config;
+
+import javax.validation.constraints.NotNull;
+
```
Comment (kinpang-CH) [inline src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java:4]: Add Javadocs for this class.

File: src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLinksWrapper.java [hunk 6]
```diff
@@ -1,33 +1,70 @@
...
+import java.util.Comparator;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Aggregates {@link ResourceLink download links} to comparison results.
  */
 public class ResourceLinksWrapper {
 
     private final List<ResourceLink> downloadLinkList;
```
Comment (kinpang-CH) [inline src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLinksWrapper.java:13]: Change to set, and construct a Treeset at runtime.

File: src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java [hunk 7]
```diff
@@ -8,50 +8,124 @@
...
+        newExchange.getIn().setHeader("ResourceLinkDescription", "Description2");
+        newExchange.getIn().setHeader("ComparisonGroup", "Company Profile");
+        newExchange.getIn().setHeader("LinkId", "link-id-2");
+
+        //when
+        when(aggregationHandler.getAggregationConfiguration(anyString())).thenReturn(comparisonGroupModel);
+        when(comparisonGroupModel.getEmailLinkModel()).thenReturn(emailLinkModelMap);
+        when(emailLinkModelMap.get("link-id-2")).thenReturn(emailLinkModel);
+        when(emailLinkModel.getRank()).thenReturn((short)20);
+
```
Comment (kinpang-CH) [inline src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java:101]: Move above lines to given section.

File: src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java [hunk 8]
```diff
@@ -8,50 +8,124 @@
...
         exchange.getIn().setHeader("ResourceLinkDescription", "Description");
+        exchange.getIn().setHeader("ComparisonGroup", "any");
+        exchange.getIn().setHeader("LinkId", "any");
 
         //when
+        when(aggregationHandler.getAggregationConfiguration(anyString())).thenReturn(comparisonGroupModel);
+        when(comparisonGroupModel.getEmailLinkModel()).thenReturn(emailLinkModelMap);
+        when(emailLinkModelMap.get(anyString())).thenReturn(emailLinkModel);
+        when(emailLinkModel.getRank()).thenReturn((short)10);
+
```
Comment (kinpang-CH) [inline src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java:128]: Move above lines to given section.

File: src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/S3EmailPublisherAggregationStrategyTest.java [hunk 9]
```diff
@@ -95,17 +101,34 @@ void testAggregateFurtherMessages() {
...
         assertNull(actual.getIn().getHeader("Completed"));
         verify(aggregationHandler).getAggregationConfiguration("group");
     }
 
+    @Test
+    void testIllegalArgumentExceptionIsThrowIfComparisonGroupModelAbsent() {
+        // given
+        //Exchange prev = new DefaultExchange(context);
+        //prev.getIn().setHeader("PublisherResourceRequests", new PublisherResourceRequestWrapper(new ArrayList<>(Collections.singletonList(new PublisherResourceRequest("key", 300L, "uploaderEndpoint", "presignerEndpoint", "resourceLinkDescription", "results".getBytes(), "fakegroup", "linkId")))));
+
```
Comment (kinpang-CH) [inline src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/S3EmailPublisherAggregationStrategyTest.java:115]: Remove above commented out lines.

File: src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLinksWrapper.java (hunk 6 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/reconciliation/model/ResourceLinksWrapper.java]: Resolved in latest commit.

File: src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java (hunk 8 above)
Comment (kinpang-CH) [reply src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java:128]: Resolved in latest commit.

File: src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java (hunk 5 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java:6]: Resolved in latest commit.

File: src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java (hunk 7 above)
Comment (kinpang-CH) [reply src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/EmailAggregationStrategyTest.java:101]: Resolved in latest commit.

File: src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/S3EmailPublisherAggregationStrategyTest.java (hunk 9 above)
Comment (kinpang-CH) [reply src/test/java/uk/gov/companieshouse/reconciliation/function/email/aggregator/S3EmailPublisherAggregationStrategyTest.java]: Resolved in latest commit.

File: src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java [hunk 10]
```diff
@@ -20,10 +30,18 @@ public void setGroupName(String groupName) {
...
 
+    /**
+     * @return size representing the number email link models in a comparison group
+     */
     public int getSize() {
         return emailLinkModel.size();
     }
 
+    /**
+     * Returns Map of {@link EmailLinkModel}'s indexed by emailId.
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java:41]: Please remove the apostrophe from EmailLinkModel's.

File: src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java [hunk 11]
```diff
@@ -1,7 +1,12 @@
 package uk.gov.companieshouse.reconciliation.config;
 
+import uk.gov.companieshouse.reconciliation.model.ResourceLink;
+
 import javax.validation.constraints.NotNull;
 
+/**
+ * Email link model that defines a rank that can be used to order {@link ResourceLink}'s.
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java:8]: Please remove the apostrophe from ResourceLink's.

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java (hunk 11 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/reconciliation/config/EmailLinkModel.java]: Resolved in latest commit.

File: src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java (hunk 10 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java]: Resolved in latest commit.

File: src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java [hunk 12]
```diff
@@ -12,6 +17,11 @@
     @NotNull
     private Map<String, EmailLinkModel> emailLinkModel;
 
+    /**
+     * Returns the group name corresponding to the group name of a Comparison group.
```
[the remark this answers is not in the document]
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java]: Resolved in latest commit.

File: src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java [hunk 13]
```diff
@@ -4,6 +4,11 @@
 import javax.validation.constraints.NotNull;
 import java.util.Map;
 
+/**
+ * Model that defines a logical grouping of {@link EmailLinkModel}'s.
```
[the remark this answers is not in the document]
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/reconciliation/config/ComparisonGroupModel.java]: Resolved in latest commit.

**Review by dlgroves-ch — APPROVED** [review]

**Review by cgregoryCH — APPROVED** [review]

**Pull request merged**